src/parser/ast/expression.rs

```diff
@@ -0,0 +1,215 @@
+impl BinaryOperator {
+    pub fn is_logic(&self) -> bool {
+        use BinaryOperator::*;
+        match self {
+            Eq => true,
+            NEq => true,
+            And => true,
+            Or => true,
+            Sum => false,
+            _ => false,
+        }
+    }
+
+    pub fn is_arith(&self) -> bool {
+        !self.is_logic()
+    }
+}
+
+impl Debug for BinaryOperator {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Sum => write!(f, "+"),
+            Self::Sub => write!(f, "-"),
+            Self::Mul => write!(f, "*"),
+            Self::Pow => write!(f, "^"),
+            Self::Eq => write!(f, "=="),
+            Self::NEq => write!(f, "!="),
+            Self::And => write!(f, "&&"),
+            Self::Or => write!(f, "||"),
+        }
+    }
+}
+
+impl From<String> for BinaryOperator {
+    fn from(value: String) -> Self {
+        use BinaryOperator::*;
+        match value.as_str() {
+            "+" => Sum,
+            "*" => Mul,
+            "^" => Pow,
+            "==" => Eq,
+            "!=" => NEq,
+            "&&" => Or,
+            "||" => And,
+            &_ => unimplemented!(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub enum UnaryOperator {
+    Not,
+    Neg,
+}
+
+impl UnaryOperator {
+    pub fn is_logic(&self) -> bool {
+        use UnaryOperator::*;
+        match self {
+            Not => true,
+            Neg => false,
+        }
+    }
+
+    pub fn is_arith(&self) -> bool {
+        !self.is_logic()
+    }
+}
+
+impl Debug for UnaryOperator {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use UnaryOperator::*;
+        match self {
+            Not => write!(f, "!"),
+            Neg => write!(f, "-"),
+        }
+    }
+}
+
+impl From<String> for UnaryOperator {
+    fn from(value: String) -> Self {
+        use UnaryOperator::*;
+        match value.as_str() {
+            "!" => Not,
+            "-" => Neg,
+            &_ => unimplemented!(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub enum Expression<F, V> {
+    // Infix op
+    BinOp {
+        dsym: DebugSymRef,
+        op: BinaryOperator,
+        lhs: Box<Expression<F, V>>,
+        rhs: Box<Expression<F, V>>,
+    },
+
+    UnaryOp {
+        dsym: DebugSymRef,
+        op: UnaryOperator,
+        sub: Box<Expression<F, V>>,
+    },
+
+    Select(
+        DebugSymRef,
+        Box<Expression<F, V>>,
+        Box<Expression<F, V>>,
+        Box<Expression<F, V>>,
+    ),
+
+    // Terminals
+    Query(DebugSymRef, V),
+    Const(DebugSymRef, F),
+    True(DebugSymRef),
+    False(DebugSymRef),
+}
+
+impl<F, V> Expression<F, V> {
+    pub fn is_arith(&self) -> bool {
+        use Expression::*;
+        match self {
+            BinOp { op, .. } => op.is_arith(),
+            UnaryOp { op, .. } => op.is_arith(),
+            Select(_, _, _, _) => true,
+            Query(_, _) => true,
+            Const(_, _) => true,
```

Comment:
What about boolean constants?

Reply:
There are True and False as their own AST nodes